### PR TITLE
Refactor the `InjectBindingRegistry` to perform validation 

### PR DESCRIPTION
**Pull request by cgruber**
Refactor the `InjectBindingRegistry` to perform validation on the `@Inject` constructors and members injections, so that they get validated when added during construction of the `BindingGraph`.

-------------
Created by MOE: https://github.com/google/moe
MOE_MIGRATED_REVID=111129042

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ Both versions have benefitted from strong involvement from Square, Google, and o
 Status
 ------
 
-  - ***Release Version:* 2.0.2**
+  - ***Release Version:* 2.0.1**
```
Comment (ronshapiro) [inline README.md]: Looks like this killed your README change

Comment (cgruber) [inline README.md]: Crap.  Good catch.  I'll fix.

**Comment by cgruber** [comment]
There. That should do it.

**Comment by cgruber** [comment]
Passes locally, fwiw

**Comment by gk5885** [comment]
LGTM

**Pull request merged**